server/src/test/java/org/elasticsearch/gateway/MetaStateServiceTests.java

```diff
@@ -21,83 +21,163 @@
+
+            metaStateService.writeGlobalState("write1", metaData);
+            metaStateService.writeIndex("write1", notChangedIndex);
+            metaStateService.writeIndex("write1", removedIndex);
+            metaStateService.writeIndex("write1", changedIndex_v1);
+            metaStateService.writeMetaState("write1");
+
+            metaStateService.keepIndex(notChangedIndex.getIndex());
+            metaStateService.writeIndex("write2", changedIndex_v2);
+            metaStateService.writeIndex("write2", newIndex);
+            metaStateService.writeMetaState("write2");
+
+            MetaData loadedMetaData = metaStateService.loadMetaData();
+            assertTrue(MetaData.isGlobalStateEquals(loadedMetaData, metaData));
+            assertThat(loadedMetaData.index("not_changed_index"), equalTo(notChangedIndex));
+            assertThat(loadedMetaData.index("removed_index"), is(nullValue()));
+            assertThat(loadedMetaData.index("changed_index"), equalTo(changedIndex_v2));
+            assertThat(loadedMetaData.index("new_index"), equalTo(newIndex));
         }
     }
+
+    public void testLoadZen1BWC() throws Exception {
+        try (NodeEnvironment env = newNodeEnvironment()) {
+            MetaStateService metaStateService = new MetaStateService(Settings.EMPTY, env, xContentRegistry());
+            MetaData metaData_v1 = MetaData.builder()
+                    .persistentSettings(Settings.builder().put("test1", "value1").build())
+                    .build();
+            MetaData metaData_v2 = MetaData.builder()
+                    .persistentSettings(Settings.builder().put("test1", "value2").build())
+                    .build();
+
+            IndexMetaData index_v1 = IndexMetaData.builder("index").settings(indexSettings()).build();
+            IndexMetaData index_v2 = IndexMetaData.builder(index_v1).version(index_v1.getVersion() + 1).build();
+
+            metaStateService.writeGlobalState("write1", metaData_v1);
+            metaStateService.writeIndex("write1", index_v1);
+            metaStateService.writeMetaState("write1");
+
+            metaStateService.writeGlobalState("write2", metaData_v2);
+            metaStateService.writeIndex("write2", index_v2);
+            //we don't write manifest file here
+
+            MetaData loadedMetaData = metaStateService.loadMetaData(); //this must load old metadata
+            assertTrue(MetaData.isGlobalStateEquals(loadedMetaData, metaData_v1));
+            assertThat(loadedMetaData.index("index"), equalTo(index_v1));
+
+            MetaState.FORMAT.cleanupOldFiles(Long.MAX_VALUE, env.nodeDataPaths()); // this will erase manifest file
+            loadedMetaData = metaStateService.loadMetaData(); //this must load new metadata, because manifest file is gone
+            assertTrue(MetaData.isGlobalStateEquals(loadedMetaData, metaData_v2));
+            assertThat(loadedMetaData.index("index"), equalTo(index_v2));
```

Comment:
I think we should also verify that we can correctly go from not having a manifest file to having one, because that's the sequence that we want to support.

Reply:
++

Reply:
https://github.com/elastic/elasticsearch/pull/35199/commits/ad0fcdb06744ffdb76445d2d6f73c50983b0f4e3